Userland/Libraries/LibCore/Group.cpp

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2022, Kenneth Myhra <kennethmyhra@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/Group.h>
+#include <LibCore/System.h>
+#include <ctype.h>
+
+namespace Core {
+
+ErrorOr<Optional<Group>> Group::from_name(StringView group_name)
+{
+    auto group = TRY(Core::System::getgrnam(group_name));
+    if (!group.has_value()) {
+        return Optional<Group> {};
+    }
+
+    return TRY(from_libc_group(group.value()));
+}
+
+ErrorOr<Optional<Group>> Group::from_gid(gid_t group_id)
+{
+    auto group = TRY(Core::System::getgrgid(group_id));
+    if (!group.has_value()) {
+        return Optional<Group> {};
+    }
+
+    return TRY(from_libc_group(group.value()));
+}
+
+#ifndef AK_OS_MACOS
+ErrorOr<void> Group::add_group(Group& group)
+{
+    if (group.group_name().is_null_or_empty())
+        return Error::from_string_literal("Group::group_name() is required.");
+
+    if (::strpbrk(group.group_name().characters(), "\\/!@#$%^&*()~+=`:\n"))
+        return Error::from_string_literal("Group name has invalid characters.");
+
+    if (group.group_name().characters()[0] == '_' || group.group_name().characters()[0] == '-' || !::isalpha(group.group_name().characters()[0]))
+        return Error::from_string_literal("Group name has invalid characters.");
+
+    if (TRY(Core::System::getgrnam(group.group_name())).has_value())
+        return Error::from_string_literal("Group name already exists.");
+
+    if (group.group_id() > 0) {
+        if (TRY(Core::System::getgrgid(group.group_id())).has_value())
+            return Error::from_string_literal("Group ID already exists.");
+    } else {
+        unsigned int group_id = 100;
+        while (true) {
+            auto gr = TRY(Core::System::getgrgid(group_id));
+            if (!gr.has_value())
+                break;
+            group_id++;
+        }
+        group.set_group_id(group_id);
+    }
+
+    auto gr = TRY(group.to_libc_group());
+
+    errno = 0;
+    FILE* file = ::fopen("/etc/group", "a");
+    if (errno)
+        return Error::from_errno(errno);
+
+    int rc = ::putgrent(&gr, file);
+    if (rc < 0)
+        return Error::from_errno(errno);
+
+    ::fclose(file);
+
+    return {};
+}
+#endif
+
+Group::Group(String group_name, unsigned int group_id, Vector<String> group_members)
+    : m_group_name(move(group_name))
+    , m_group_id(group_id)
+    , m_group_members(move(group_members))
+{
+}
+
+ErrorOr<struct group> Group::to_libc_group()
+{
+    struct group gr;
+    ::memset(&gr, 0, sizeof(struct group));
+    gr.gr_name = const_cast<char*>(group_name().characters());
```

Comment:
When `*this` is destroyed, `this->m_group_name` will be freed. This is a UAF.

Reply:
I do not see an issue with that. `to_libc_group()` is a private convenience method which is only called right before we write the group entry to disk.